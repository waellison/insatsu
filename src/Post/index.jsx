import React from "react";
import { useParams } from "react-router-dom"
import { marked } from "marked";
import Posts from "../Posts";
import PostLinks from "../PostLinks";
import NavBar from "../NavBar";

class PostItem extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            id: this.props.params.id,
            post: undefined,
            loaded: false
        };
        this.getMatchingPost = this.getMatchingPost.bind(this);
    }
    
    getMatchingPost() {
        return fetch(`http://localhost:5000/posts/${this.state.id}`, {mode: 'cors'});
    }

    componentDidMount() {
        this.getMatchingPost()
            .then(res => res.json())
            .then(
                (result) => {
                    this.setState({
                        post: result,
                        loaded: true
                    });
                },
                (error) => {
                    this.setState({
                        loaded: true,
                        error
                    });
                }
            );
    }

    render() {
        let {id, error, loaded, post} = this.state;
        let postMarkup = undefined;

        if(error || this.state.hasError) {
            return (
                <div className="oopsIFuckedUp">
                    {error.message}
                </div>
            );
        } else if(!loaded) {
            return (
                <div className="justASecAlrightGeez">
                    Loading...
                </div>
            );
        }

        if(id === "all") {
            return <Posts />
        } else if(post) {
            postMarkup = marked.parse(post.content)

            return (
                <article>
                    <h1>{post.name}</h1>
                    <main dangerouslySetInnerHTML={{__html: postMarkup}} />
                    <PostLinks next={post.next_id} prev={post.previous_id} />
                </article>            
            );
        }
    }
}

const Post = (props) => (
    <>
    <NavBar />
    <PostItem
        {...props}
        params={useParams()}/>  
    </>
);

export default Post;
