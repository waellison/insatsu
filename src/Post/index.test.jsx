import React from "react";
import renderer from 'react-test-renderer';
import fetchMock from "jest-fetch-mock";
import Post from './index';

fetchMock.enableMocks();

it('renders as expected when given a post to render', () => {
    const postMock = {
        "categories": [
            {
                "id": 16,
                "name": "Starsmerchant Nuts",
                "parentCategoryId": null,
                "slug": "starsmerchant-nuts",
                "summary": "Practice bar budget to always. According order how nation.",
                "taxonomyType": "category"
            }
        ],
        "content": "Line option improve trip. Try our throughout drop three even late.\n\nLow start trouble subject significant threat nature. Training then own.\n\nLarge history husband. Down summer wind experience have. Truth wear some city.\n\nScene until already. Late dog lay something practice.\n\nVoice he effort information military soldier. Republican scene three art institution image most organization. Million report there child.\n\nWould instead too peace her general scene. Difficult pressure positive head research last.\n\nNot word reflect despite speech range series. Interview current nice I fly human. Prevent hour house rather design. When coach green short really create major.\n\nLand single him way value network tonight. Collection mean already conference. Everyone by have figure thus may pick.\n\nParticular you management force population. Receive time personal hair yes feel.\n\nTonight recent site analysis could investment street bar. Culture century game. War if drive minute garden into ask.",
        "id": 1,
        "is_published": true,
        "name": "Fare Concrete Hill Apparatus Imported",
        "next_id": 2,
        "previous_id": null,
        "slug": "fare-concrete-hill-apparatus-imported",
        "summary": "Mouth stuff statement thousand economic should. Majority nature back mean.",
        "tags": []
    };

    jest.mock('./index.jsx');
    fetch.mockResponseOnce(JSON.stringify(postMock));

    const post = renderer.create(<Post />);
    let tree = post.toJSON();
    expect(tree).toMatchSnapshot();
    expect(fetch).toHaveBeenCalledTimes(1);
});

it('renders as expected upon error', () => {
    fetch.mockReject(() => Promise.reject("404 not found"));

    const post = renderer.create(<Post />);
    // do not set state here
    let tree = post.toJSON();
    expect(tree).toMatchSnapshot();
});
