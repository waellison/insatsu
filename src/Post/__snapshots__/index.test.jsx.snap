// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected upon error 1`] = `
<div
  className="justASecAlrightGeez"
>
  Loading...
</div>
`;

exports[`renders as expected when given a post to render 1`] = `
<div
  className="justASecAlrightGeez"
>
  Loading...
</div>
`;
